cmd/utils.go

```diff
@@ -693,18 +694,47 @@ func ceilFrac(numerator, denominator int64) (ceil int64) {
 	return
 }
 
-// Returns context with ReqInfo details set in the context.
-func newContext(r *http.Request, w http.ResponseWriter, api string) context.Context {
-	vars := mux.Vars(r)
-	bucket := vars["bucket"]
-	object, err := url.PathUnescape(vars["object"])
+// unescapeGeneric is similar to url.PathUnescape or url.QueryUnescape
+// depending on input, additionally also handles situations such as
+// `//` are normalized as `/`, also removes any `/` prefix before
+// returning.
+func unescapeGeneric(p string, escapeFn func(string) (string, error)) (string, error) {
+	ep, err := escapeFn(p)
 	if err != nil {
-		object = vars["object"]
+		ep = p
 	}
-	prefix, err := url.QueryUnescape(vars["prefix"])
+	if i := strings.Index(ep, SlashSeparator); i == 0 {
```

Comment:
No need to search the entire string when we only want to check the first char:
```suggestion
	if len(ep) > 0 && ep[0] == SlashSeparator {
```